src/aws-cpp-sdk-core/source/utils/crypto/bcrypt/CryptoImpl.cpp

```diff
@@ -86,34 +86,58 @@ namespace Aws
             {
             public:
 
-                BCryptHashContext(void* algorithmHandle, PBYTE hashObject, DWORD hashObjectLength) :
+                BCryptHashContext(void* algorithmHandle, DWORD hashObjectLength) :
                     m_hashHandle(nullptr),
-                    m_isValid(false)
+                    m_isValid(false),
+                    m_hashBufferLength(hashObjectLength),
+                    m_hashObject(nullptr)
                 {
-                    NTSTATUS status = BCryptCreateHash(algorithmHandle, &m_hashHandle, hashObject, hashObjectLength, nullptr, 0, 0);
+                    m_hashObject = Aws::NewArray<BYTE>(m_hashObjectLength, logTag);
+                    if (!m_hashObject)
+                    {
+                        AWS_LOGSTREAM_ERROR(logTag, "Error allocating hash object.");
+                        return;
+                    }
+
+                    NTSTATUS status = BCryptCreateHash(algorithmHandle, &m_hashHandle, m_hashObject, m_hashObjectLength, nullptr, 0, 0);
                     m_isValid = NT_SUCCESS(status);
                 }
 
-                BCryptHashContext(void* algorithmHandle, PBYTE hashObject, DWORD hashObjectLength, const ByteBuffer& secret) :
+                BCryptHashContext(void* algorithmHandle, DWORD hashObjectLength, const ByteBuffer& secret) :
                     m_hashHandle(nullptr),
-                    m_isValid(false)
+                    m_isValid(false),
+                    m_hashBufferLength(hashObjectLength),
+                    m_hashObject(nullptr)
                 {
-                    NTSTATUS status = BCryptCreateHash(algorithmHandle, &m_hashHandle, hashObject, hashObjectLength, secret.GetUnderlyingData(), (ULONG)secret.GetLength(), 0);
+                    m_hashObject = Aws::NewArray<BYTE>(m_hashObjectLength, logTag);
+                    if (!m_hashObject)
+                    {
+                        AWS_LOGSTREAM_ERROR(logTag, "Error allocating hash object.");
+                        return;
+                    }
+
+                    NTSTATUS status = BCryptCreateHash(algorithmHandle, &m_hashHandle, m_hashObject, m_hashObjectLength, secret.GetUnderlyingData(), (ULONG)secret.GetLength(), 0);
                     m_isValid = NT_SUCCESS(status);
                 }
 
                 ~BCryptHashContext()
                 {
+                    // Destruct in inverted order
                     if (m_hashHandle)
                     {
                         BCryptDestroyHash(m_hashHandle);
+                        m_hashHandle = nullptr;
                     }
+
+                    Aws::DeleteArray(m_hashObject);
                 }
 
                 bool IsValid() const { return m_isValid; }
 
                 BCRYPT_HASH_HANDLE m_hashHandle;
                 bool m_isValid;
+                DWORD m_hashObjectLength;
```

Comment:
hello, thanks a lot for this great finding
it seems that you have forgotten to add `m_hashBufferLength` here
(or more likely have accidentally added `m_hashBufferLength` to the c-tor member init list).

Reply:
> hello, thanks a lot for this great finding it seems that you have forgotten to add `m_hashBufferLength` here (or more likely have accidentally added `m_hashBufferLength` to the c-tor member init list).

yes, there is a typo there, should be `m_hashObjectLength`, fixed that in the newest commit. Btw, the caller of BCryptHashContext::BCryptHashContext() had missed to remove redundant parameters, also fixed that.